File: src/tbmcp/tools/folders.py
"""The folder tree: browsing it, reshaping it, and emptying it out.

`folder_list` is the entry point for most other tools — folder ids are opaque and
only valid while Thunderbird stays open, so anything that needs one comes here
first. Its output is deliberately terse for that reason.
"""

# NOTE: no `from __future__ import annotations` in toolset modules. Tool signatures
# must evaluate at definition time so `Gate(...)` produces a real
# `Annotated[Consent, Resolve(...)]` rather than a string the SDK has to re-evaluate.

from typing import Any, Literal

from ..errors import UsageError
from ..safety import (
    DESTRUCTIVE,
    IDEMPOTENT_WRITE,
    MUTATING,
    Gate,
    guard_write,
    require,
)
from ..server import Registrar
from ._common import call, changed, clamp, dry_run, page

SpecialUse = Literal["inbox", "drafts", "sent", "trash", "templates", "archives", "junk"]

#: Depth 1 is account roots plus their top-level folders — enough to orient a model
#: without paying for a 400-folder IMAP tree it did not ask for.
DEFAULT_DEPTH = 1
MAX_DEPTH = 20


def _slim(folder: dict[str, Any]) -> dict[str, Any]:
    """Drop the fields that carry no information, and recurse into `children`.

    `folder_list` runs on nearly every conversation, so a key that is always null
    or always empty is a key not worth spending the model's context on.
    """
    out = {
        key: value
        for key, value in folder.items()
        if key != "children" and value is not None and value != [] and value != ""
    }
    children = folder.get("children")
    if children:
        out["children"] = [_slim(child) for child in children]
    return out


def register(reg: Registrar) -> None:
    # ------------------------------------------------------------------- browsing

    @reg.read_tool(title="List folders")
    async def folder_list(
        account_id: str | None = None,
        name: str | None = None,
        parent_id: str | None = None,
        special_use: SpecialUse | None = None,
        is_virtual: bool | None = None,
        is_tag: bool | None = None,
        is_unified: bool | None = None,
        is_favorite: bool | None = None,
        is_root: bool | None = None,
        has_unread_messages: bool | None = None,
        depth: int | None = None,
        tree: bool = False,
        include_counts: bool = True,
        limit: int = 200,
    ) -> dict[str, Any]:
        """List mail folders with their ids and message counts.

        With no filters this browses: every account's root folder plus one level
        below it. Raise `depth` to walk further, or give a filter — `name` is a
        case-insensitive substring — and the whole tree is searched instead. Pass
        `tree=true` to get folders nested under `children` rather than flat.

        Counts come from Thunderbird's own folder database. On IMAP that database
        can lag until the folder has been selected once in Thunderbird, so treat
        unread and total counts as close rather than exact.
        """
        query: dict[str, Any] = {}
        if account_id:
            query["accountId"] = account_id
        if name:
            query["name"] = name
        if parent_id:
            query["folderId"] = parent_id
            query["recursive"] = True
        if special_use:
            query["specialUse"] = [special_use]
        for key, value in (
            ("isVirtual", is_virtual),
            ("isTag", is_tag),
            ("isUnified", is_unified),
            ("isFavorite", is_favorite),
            ("isRoot", is_root),
            ("hasUnreadMessages", has_unread_messages),
        ):
            if value is not None:
                query[key] = value

        if depth is not None:
            depth = clamp(depth, default=DEFAULT_DEPTH, minimum=0, maximum=MAX_DEPTH, field="depth")
        result = await call(
            "folders.query",
            {
                "query": query,
                "depth": depth,
                "tree": tree,
                "includeCounts": include_counts,
                "limit": clamp(limit, default=200, minimum=1, maximum=500, field="limit"),
            },
            timeout=90.0,
        )
        return page(
            [_slim(folder) for folder in result.get("folders", [])],
            truncated=bool(result.get("truncated")),
            searchedWholeTree=result.get("searched"),
        )

    @reg.read_tool(title="Get a folder")
    async def folder_get(folder_id: str, include_subfolders: bool = False) -> dict[str, Any]:
        """Get one folder: counts, special use, flags and IMAP quota.

        Use this to re-check a count after a move or a delete; `folder_list` is the
        cheaper way to find the id in the first place.
        """
        result = await call(
            "folders.get",
            {"folderId": folder_id, "includeSubFolders": include_subfolders},
            timeout=60.0,
        )
        return _slim(result.get("folder") or {})

    @reg.read_tool(title="Get folder capabilities")
    async def folder_capabilities(folder_id: str) -> dict[str, Any]:
        """Report what may be done to a folder before attempting it.

        Answers whether the folder can hold messages, take subfolders, be renamed,
        be deleted, or have messages deleted from it. Worth a call before offering
        the user a plan that a server would refuse.
        """
        result = await call("folders.capabilities", {"folderId": folder_id}, timeout=30.0)
        return {"folderId": folder_id, **(result.get("capabilities") or {})}

    @reg.read_tool(title="Get a unified folder")
    async def folder_get_unified(
        folder_type: SpecialUse, include_subfolders: bool = False
    ) -> dict[str, Any]:
        """Get the unified folder that spans every account, e.g. all inboxes at once.

        Its id works anywhere a folder id is accepted, so `mail_list` on the unified
        inbox lists new mail across all accounts in one call.
        """
        result = await call(
            "folders.getUnified",
            {"type": folder_type, "includeSubFolders": include_subfolders},
            timeout=60.0,
        )
        return _slim(result.get("folder") or {})

    # ------------------------------------------------------------------- reshaping

    @reg.write_tool(title="Create a folder", annotations=MUTATING)
    async def folder_create(
        name: str,
        parent_id: str | None = None,
        account_id: str | None = None,
        confirm: bool = False,
        consent: Gate("create a folder") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Create a folder inside another folder, or at the top of an account."""
        guard_write("create folders")
        if not parent_id and not account_id:
            raise UsageError(
                "Give parent_id to nest the folder inside another, or account_id to "
                "create it at the top level of an account."
            )
        require(consent, "create this folder")
        result = await call(
            "folders.create",
            {"name": name, "parentId": parent_id, "accountId": account_id},
            timeout=60.0,
        )
        return changed("folders.create", before=None, after=_slim(result.get("folder") or {}))

    @reg.write_tool(title="Rename a folder", annotations=MUTATING)
    async def folder_rename(
        folder_id: str,
        new_name: str,
        confirm: bool = False,
        consent: Gate("rename a folder") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Rename a folder, keeping its messages and subfolders."""
        guard_write("rename folders")
        require(consent, "rename this folder")
        result = await call(
            "folders.rename", {"folderId": folder_id, "newName": new_name}, timeout=120.0
        )
        return changed(
            "folders.rename",
            before=_slim(result.get("previous") or {}),
            after=_slim(result.get("folder") or {}),
            note="A folder's id encodes its path, so it changes on rename. Use the new id.",
        )

    @reg.write_tool(title="Move a folder", annotations=MUTATING)
    async def folder_move(
        folder_id: str,
        destination_id: str,
        confirm: bool = False,
        consent: Gate("move a folder and everything in it") = None,  # type: ignore[valid-type]
        dry_run_only: bool = False,
    ) -> dict[str, Any]:
        """Move a folder under a different parent, with its subfolders.

        `destination_id` is the new parent folder — use an account's root folder to
        move it to the top level. Across accounts this copies then deletes, which on
        IMAP can take a while.
        """
        guard_write("move folders")
        if dry_run_only:
            return dry_run("folders.move", {"folderId": folder_id, "destinationId": destination_id})
        require(consent, "move this folder")
        result = await call(
            "folders.move",
            {"folderId": folder_id, "destinationId": destination_id},
            timeout=300.0,
        )
        return changed(
            "folders.move",
            before=_slim(result.get("previous") or {}),
            after=_slim(result.get("folder") or {}),
            note="Folder and message ids below the moved folder all change; re-query.",
        )

    @reg.write_tool(title="Copy a folder", annotations=MUTATING)
    async def folder_copy(
        folder_id: str,
        destination_id: str,
        confirm: bool = False,
        consent: Gate("copy a folder and everything in it") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Copy a folder and its contents under another parent, leaving the original."""
        guard_write("copy folders")
        require(consent, "copy this folder")
        result = await call(
            "folders.copy",
            {"folderId": folder_id, "destinationId": destination_id},
            timeout=300.0,
        )
        return {
            "copied": True,
            "sourceFolderId": result.get("sourceFolderId", folder_id),
            "folder": _slim(result.get("folder") or {}),
        }

    @reg.write_tool(title="Delete a folder", annotations=DESTRUCTIVE)
    async def folder_delete(
        folder_id: str,
        confirm: bool = False,
        consent: Gate("delete a folder and everything in it") = None,  # type: ignore[valid-type]
        dry_run_only: bool = False,
    ) -> dict[str, Any]:
        """Delete a folder, its subfolders and every message in them.

        Thunderbird moves the folder to Trash unless it is already inside Trash, in
        which case it goes for good. Check `folder_get` first if the count matters —
        the reply reports what was removed, but cannot put it back.
        """
        guard_write("delete folders")
        if dry_run_only:
            return dry_run("folders.delete", {"folderId": folder_id})
        require(consent, "delete this folder")
        result = await call("folders.delete", {"folderId": folder_id}, timeout=300.0)
        permanent = bool(result.get("permanent"))
        return changed(
            "folders.delete",
            before=_slim(result.get("deleted") or {}),
            after=None,
            subFolderCount=result.get("subFolderCount"),
            permanent=permanent,
            recoverable=not permanent,
        )

    # -------------------------------------------------------------------- triaging

    @reg.write_tool(title="Mark a folder read", annotations=IDEMPOTENT_WRITE)
    async def folder_mark_read(
        folder_id: str,
        include_subfolders: bool = False,
        confirm: bool = False,
        consent: Gate("mark a whole folder as read") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Mark every message in a folder as read.

        There is no per-message undo for this, which is why it asks first. On IMAP
        the flags are pushed to the server.
        """
        guard_write("mark folders as read")
        require(consent, "mark this folder as read")
        result = await call(
            "folders.markAsRead",
            {"folderId": folder_id, "includeSubFolders": include_subfolders},
            timeout=180.0,
        )
        return changed(
            "folders.markAsRead",
            before={"unreadMessageCount": result.get("previousUnread")},
            after={"unreadMessageCount": result.get("unread")},
            folderId=folder_id,
            foldersAffected=result.get("folders"),
            failures=result.get("failures") or [],
        )

    @reg.write_tool(title="Favourite a folder", annotations=IDEMPOTENT_WRITE, interactive=False)
    async def folder_set_favorite(folder_id: str, favorite: bool = True) -> dict[str, Any]:
        """Add or remove a folder from the user's favourites.

        Cosmetic and reversible — it only affects the folder pane's Favourites view,
        so it is not gated. Prompting for something this harmless would only train
        the user to click through the prompts that do matter.
        """
        guard_write("change folder favourites")
        result = await call(
            "folders.update", {"folderId": folder_id, "isFavorite": favorite}, timeout=30.0
        )
        return changed(
            "folders.update",
            before=_slim(result.get("previous") or {}),
            after=_slim(result.get("folder") or {}),
        )

    # ---------------------------------------------------------------- emptying out

    @reg.write_tool(title="Empty Trash", annotations=DESTRUCTIVE)
    async def folder_empty_trash(
        account_id: str | None = None,
        folder_id: str | None = None,
        remove_subfolders: bool = True,
        confirm: bool = False,
        consent: Gate("permanently delete everything in Trash") = None,  # type: ignore[valid-type]
        dry_run_only: bool = False,
    ) -> dict[str, Any]:
        """Permanently delete everything in one account's Trash.

        This is not recoverable from Thunderbird. Give `account_id` and the account's
        configured Trash is used; give `folder_id` to empty a specific one. Subfolders
        of Trash are removed too unless `remove_subfolders=false`.
        """
        guard_write("empty Trash")
        if not account_id and not folder_id:
            raise UsageError(
                "Give account_id (its configured Trash) or folder_id (a specific Trash "
                "folder) — emptying every account's Trash at once is not offered."
            )
        target = {
            "accountId": account_id,
            "folderId": folder_id,
            "removeSubFolders": remove_subfolders,
        }
        if dry_run_only:
            return dry_run("folders.emptyTrash", target)
        require(consent, "permanently delete everything in Trash")
        result = await call("folders.emptyTrash", target, timeout=600.0)
        return changed(
            "folders.emptyTrash",
            before={"path": result.get("path"), "messages": result.get("deleted")},
            after={"messages": 0},
            deleted=result.get("deleted"),
            foldersRemoved=result.get("foldersRemoved"),
            recoverable=False,
        )

    @reg.write_tool(title="Empty Junk", annotations=DESTRUCTIVE)
    async def folder_empty_junk(
        account_id: str | None = None,
        folder_id: str | None = None,
        confirm: bool = False,
        consent: Gate("permanently delete everything in Junk") = None,  # type: ignore[valid-type]
        dry_run_only: bool = False,
    ) -> dict[str, Any]:
        """Permanently delete everything in one account's Junk folder.

        Subfolders of Junk are emptied but kept, since they are usually filter
        targets the user set up deliberately. Not recoverable.
        """
        guard_write("empty Junk")
        if not account_id and not folder_id:
            raise UsageError(
                "Give account_id (its configured Junk folder) or folder_id (a specific one)."
            )
        target = {"accountId": account_id, "folderId": folder_id}
        if dry_run_only:
            return dry_run("folders.emptyJunk", target)
        require(consent, "permanently delete everything in Junk")
        result = await call("folders.emptyJunk", target, timeout=600.0)
        return changed(
            "folders.emptyJunk",
            before={"path": result.get("path"), "messages": result.get("deleted")},
            after={"messages": 0},
            deleted=result.get("deleted"),
            subFoldersEmptied=result.get("subFoldersEmptied"),
            recoverable=False,
        )

    # ---------------------------------------------------------------- housekeeping

    @reg.write_tool(title="Download a folder for offline use", annotations=MUTATING)
    async def folder_sync_offline(
        folder_id: str,
        include_subfolders: bool = False,
        confirm: bool = False,
        consent: Gate("download a folder for offline use") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Fetch an IMAP folder's message bodies so they are available offline.

        This is what `mail_get_source` needs before it can return raw source for an
        IMAP message. A large folder means a long download and real network traffic.
        """
        guard_write("download folders for offline use")
        require(consent, "download this folder for offline use")
        result = await call(
            "x.admin.downloadForOffline",
            {"folderId": folder_id, "includeSubFolders": include_subfolders},
            timeout=600.0,
        )
        return {"folderId": folder_id, "started": True, **(result or {})}

    @reg.write_tool(title="Compact folders", annotations=MUTATING)
    async def folder_compact(
        folder_id: str | None = None,
        account_id: str | None = None,
        confirm: bool = False,
        consent: Gate("compact folders and rewrite their message stores") = None,  # type: ignore[valid-type]
    ) -> dict[str, Any]:
        """Reclaim the disk space left behind by deleted messages.

        Deleted mail stays in the message store until the folder is compacted. Pass
        `folder_id` for one folder or `account_id` for all of an account's. Nothing
        readable is lost, but the store is rewritten, so do not interrupt it.
        """
        guard_write("compact folders")
        if not folder_id and not account_id:
            raise UsageError("Give folder_id for one folder, or account_id for a whole account.")
        require(consent, "compact these folders")
        result = await call(
            "x.admin.compactFolders",
            {"folderId": folder_id, "accountId": account_id},
            timeout=600.0,
        )
        return {"compacted": True, **(result or {})}

File: addon/background/handlers/folders.js
/* Folder handlers — the official browser.folders / browser.accounts surface.
 *
 * `folders.query` carries most of the traffic, because a model needs a folder id
 * before it can do anything else. All the shaping lives here: one compact record
 * per folder, MailFolderInfo counts folded in, flat or nested on request.
 *
 * Three things learned against 153:
 *   - queryInfo.name compares a bare string for equality, so a substring search has
 *     to travel as a RegularExpressionType.
 *   - MailFolder.type is gone in favour of specialUse; we keep a single-word `type`
 *     because that is what callers actually branch on.
 *   - there is no emptyTrash/emptyJunk in the official API at all. See the bottom.
 */

{
  const DEFAULT_DEPTH = 1;
  const MAX_DEPTH = 20;
  const MAX_FOLDERS = 500;
  const INFO_CONCURRENCY = 8;
  const PURGE_CHUNK = 200;

  const UNIFIED_TYPES = ["inbox", "drafts", "sent", "trash", "templates", "archives", "junk"];

  /* Predicates that mean "search the whole tree". Anything else (accountId, isRoot)
   * is still just browsing, so the depth cap stays in force for it. */
  const SEARCH_KEYS = [
    "name",
    "path",
    "specialUse",
    "folderId",
    "capabilities",
    "isVirtual",
    "isTag",
    "isUnified",
    "isFavorite",
    "hasMessages",
    "hasUnreadMessages",
    "hasNewMessages",
    "hasSubFolders",
  ];

  function escapeRegExp(text) {
    return String(text).replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  /** Depth relative to the account root: "/" is 0, "/Inbox" is 1. */
  function depthOf(folder) {
    return String(folder.path || "/")
      .split("/")
      .filter(Boolean).length;
  }

  /** queryInfo for browser.folders.query, straight from the daemon's params. */
  function buildQuery(input) {
    const query = {};
    for (const [key, value] of Object.entries(input || {})) {
      if (value === null || value === undefined || value === "") {
        continue;
      }
      query[key] = value;
    }
    if (typeof query.name === "string") {
      query.name = { regexp: escapeRegExp(query.name), flags: "i" };
    }
    return query;
  }

  /** The compact folder record the Python layer hands to the model. */
  function summary(folder, info) {
    const out = {
      id: folder.id,
      name: folder.name,
      path: folder.path,
      accountId: folder.accountId,
      type: (folder.specialUse && folder.specialUse[0]) || folder.type || null,
      specialUse: folder.specialUse || [],
      depth: depthOf(folder),
    };
    // Flags are emitted only when true: a folder listing is mostly false booleans.
    for (const flag of ["isRoot", "isVirtual", "isTag", "isUnified", "isFavorite"]) {
      if (folder[flag]) {
        out[flag] = true;
      }
    }
    if (info) {
      out.totalMessageCount = info.totalMessageCount;
      out.unreadMessageCount = info.unreadMessageCount;
      out.newMessageCount = info.newMessageCount;
      if (info.favorite) {
        out.isFavorite = true;
      }
      if (info.lastUsed) {
        // lastUsed crosses the API boundary as a Date on 153 and as a timestamp on
        // older builds; both survive this.
        const stamp = new Date(info.lastUsed);
        if (!Number.isNaN(stamp.getTime())) {
          out.lastUsed = stamp.toISOString();
        }
      }
      if (info.quota && info.quota.length) {
        out.quota = info.quota;
      }
    }
    return out;
  }

  /** A MailFolder plus whatever subFolders it already carries, with no extra calls. */
  function shallow(folder) {
    const node = summary(folder, null);
    if (folder.subFolders && folder.subFolders.length) {
      node.children = folder.subFolders.map(shallow);
    }
    return node;
  }

  function flatten(folders, out) {
    for (const folder of folders || []) {
      out.push(folder);
      flatten(folder.subFolders, out);
    }
    return out;
  }

  async function withInfo(folders) {
    const results = await tbxUtil.mapLimited(folders, INFO_CONCURRENCY, (folder) =>
      // Root folders hold no mail, so getFolderInfo on them is a wasted hop.
      folder.isRoot ? null : browser.folders.getFolderInfo(folder.id)
    );
    return folders.map((folder, index) => {
      const entry = results[index];
      return summary(folder, entry && entry.ok ? entry.value : null);
    });
  }

  /** Account roots plus `depth` levels below them, breadth first. */
  async function walkTree(query, depth) {
    const rootQuery = { isRoot: true };
    if (query.accountId) {
      rootQuery.accountId = query.accountId;
    }
    let level = (await browser.folders.query(rootQuery)) || [];
    if (!level.length) {
      // Cheap insurance: if a build ever stops returning roots from query, the
      // account list still has them.
      const accounts = (await browser.accounts.list(false)) || [];
      level = accounts
        .filter((a) => !query.accountId || a.id === query.accountId)
        .map((a) => a.rootFolder)
        .filter(Boolean);
    }
    const collected = [...level];
    for (let step = 0; step < depth && level.length && collected.length < MAX_FOLDERS; step += 1) {
      const batches = await tbxUtil.mapLimited(level, INFO_CONCURRENCY, (folder) =>
        browser.folders.getSubFolders(folder.id, false)
      );
      const next = [];
      for (const children of tbxUtil.partition(batches).values) {
        next.push(...(children || []));
      }
      collected.push(...next);
      level = next;
    }
    return collected;
  }

  function byAccountThenPath(a, b) {
    if (a.accountId !== b.accountId) {
      return a.accountId < b.accountId ? -1 : 1;
    }
    if (a.path === b.path) {
      return 0;
    }
    return a.path < b.path ? -1 : 1;
  }

  /** Nest a flat list by path. A folder whose parent is missing from the set stays
   *  at the top level, so a filtered query still returns something navigable. */
  function nest(items) {
    const byKey = new Map(items.map((item) => [`${item.accountId}${item.path}`, item]));
    const roots = [];
    for (const item of items) {
      const path = item.path || "/";
      const cut = path.lastIndexOf("/");
      const parent = path === "/" ? null : byKey.get(`${item.accountId}${cut > 0 ? path.slice(0, cut) : "/"}`);
      if (parent && parent !== item) {
        parent.children = parent.children || [];
        parent.children.push(item);
      } else {
        roots.push(item);
      }
    }
    return roots;
  }

  /** The account's folder for a specialUse ("trash", "junk", …). */
  async function specialFolder(accountId, use, required) {
    const matches = (await browser.folders.query({ accountId, specialUse: [use] })) || [];
    if (matches.length) {
      return matches[0];
    }
    if (!required) {
      return null;
    }
    throw tbxError.usage(
      `account ${accountId} has no ${use} folder configured — check Account Settings ` +
        "> Copies & Folders, or pass folderId explicitly"
    );
  }

  /** Folder ids on 153, MailFolder objects on 128. Try ids (cheaper) and fall back
   *  once, rather than branching on a version number. */
  async function callWithFolders(name, ids, tail) {
    try {
      return await browser.folders[name](...ids, ...tail);
    } catch (ex) {
      const message = String(ex.message || ex);
      if (!/MailFolder|type error|Expected|does not match/i.test(message)) {
        throw ex;
      }
      const objects = await Promise.all(ids.map((id) => browser.folders.get(id)));
      return browser.folders[name](...objects, ...tail);
    }
  }

  // --------------------------------------------------------------------- reading

  tbxRegistry.define("folders.query", async (params) => {
    const query = buildQuery(params.query);
    const limit = tbxUtil.limit(params.limit, 200, MAX_FOLDERS);
    const searching = SEARCH_KEYS.some((key) => key in query);
    const depth = Number.isInteger(params.depth)
      ? Math.min(Math.max(params.depth, 0), MAX_DEPTH)
      : null;

    let folders;
    if (searching) {
      folders = (await browser.folders.query(query)) || [];
      if (depth !== null) {
        folders = folders.filter((folder) => depthOf(folder) <= depth);
      }
    } else {
      folders = await walkTree(query, depth === null ? DEFAULT_DEPTH : depth);
    }

    // Sort before truncating, so dropping the tail is predictable rather than
    // whatever order Thunderbird happened to walk the accounts in.
    folders.sort(byAccountThenPath);
    const truncated = folders.length > limit;
    folders = folders.slice(0, limit);

    const items =
      params.includeCounts === false
        ? folders.map((folder) => summary(folder, null))
        : await withInfo(folders);
    return {
      folders: params.tree ? nest(items) : items,
      count: items.length,
      truncated,
      searched: searching,
    };
  });

  tbxRegistry.define("folders.get", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const includeSubFolders = Boolean(params.includeSubFolders);
    const folder = await browser.folders.get(folderId, includeSubFolders);
    const info = folder.isRoot
      ? null
      : await browser.folders.getFolderInfo(folderId).catch(() => null);
    const payload = summary(folder, info);
    if (includeSubFolders && folder.subFolders && folder.subFolders.length) {
      payload.children = folder.subFolders.map(shallow);
    }
    return { folder: payload };
  });

  tbxRegistry.define("folders.capabilities", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    if (!browser.folders.getFolderCapabilities) {
      throw tbxError.unsupported(
        "this Thunderbird build has no folders.getFolderCapabilities (needs 128 or newer)"
      );
    }
    const capabilities = await browser.folders.getFolderCapabilities(folderId);
    return { folderId, capabilities: capabilities || {} };
  });

  tbxRegistry.define("folders.getUnified", async (params) => {
    const type = tbxUtil.need(params, "type", "string");
    if (!UNIFIED_TYPES.includes(type)) {
      throw tbxError.usage(`type must be one of ${UNIFIED_TYPES.join(", ")}`);
    }
    if (!browser.folders.getUnifiedFolder) {
      throw tbxError.unsupported(
        "this Thunderbird build has no unified folders (needs 128 or newer)"
      );
    }
    const includeSubFolders = Boolean(params.includeSubFolders);
    let folder;
    try {
      folder = await browser.folders.getUnifiedFolder(type, includeSubFolders);
    } catch (ex) {
      // Thunderbird throws rather than returning null when the unified folder for a
      // type has never been materialised on this profile.
      throw tbxError.unsupported(
        `there is no unified ${type} folder on this profile: ${ex.message || ex}`
      );
    }
    const info = await browser.folders.getFolderInfo(folder.id).catch(() => null);
    const payload = summary(folder, info);
    if (includeSubFolders && folder.subFolders && folder.subFolders.length) {
      payload.children = folder.subFolders.map(shallow);
    }
    return { folder: payload };
  });

  // -------------------------------------------------------------------- mutation

  /** A parent folder id, from parentId or from an account's root folder. */
  async function resolveParent(params) {
    if (params.parentId) {
      return params.parentId;
    }
    const accountId = tbxUtil.need(params, "accountId", "string");
    const account = await browser.accounts.get(accountId, false);
    if (!account || !account.rootFolder) {
      throw tbxError.usage(`no account with id ${accountId}`);
    }
    return account.rootFolder.id;
  }

  tbxRegistry.define("folders.create", async (params) => {
    const name = tbxUtil.need(params, "name", "string");
    const parentId = await resolveParent(params);
    const folder = await callWithFolders("create", [parentId], [name]);
    return { folder: summary(folder, null), parentId };
  });

  tbxRegistry.define("folders.rename", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const newName = tbxUtil.need(params, "newName", "string");
    const before = await browser.folders.get(folderId);
    const folder = await callWithFolders("rename", [folderId], [newName]);
    return { previous: summary(before, null), folder: summary(folder, null) };
  });

  tbxRegistry.define("folders.move", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const destinationId = tbxUtil.need(params, "destinationId", "string");
    const before = await browser.folders.get(folderId);
    const folder = await callWithFolders("move", [folderId, destinationId], []);
    return { previous: summary(before, null), folder: summary(folder, null) };
  });

  tbxRegistry.define("folders.copy", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const destinationId = tbxUtil.need(params, "destinationId", "string");
    const folder = await callWithFolders("copy", [folderId, destinationId], []);
    return { folder: summary(folder, null), sourceFolderId: folderId };
  });

  tbxRegistry.define("folders.delete", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const before = await browser.folders.get(folderId, true);
    const info = before.isRoot
      ? null
      : await browser.folders.getFolderInfo(folderId).catch(() => null);
    // Thunderbird moves a folder to Trash unless it already lives there, so work out
    // which of the two happened instead of leaving the caller to guess.
    const trash = await specialFolder(before.accountId, "trash", false).catch(() => null);
    const permanent = Boolean(
      trash && (before.path === trash.path || String(before.path).startsWith(`${trash.path}/`))
    );
    await browser.folders.delete(folderId);
    return {
      deleted: summary(before, info),
      subFolderCount: flatten(before.subFolders, []).length,
      permanent,
    };
  });

  tbxRegistry.define("folders.markAsRead", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    const before = await browser.folders.getFolderInfo(folderId).catch(() => null);
    const targets = [folderId];
    if (params.includeSubFolders) {
      const children = await browser.folders.getSubFolders(folderId, true);
      for (const folder of flatten(children, [])) {
        targets.push(folder.id);
      }
    }
    const results = await tbxUtil.mapLimited(targets, INFO_CONCURRENCY, (id) =>
      browser.folders.markAsRead(id)
    );
    const { values, failures } = tbxUtil.partition(results);
    const after = await browser.folders.getFolderInfo(folderId).catch(() => null);
    return {
      folderId,
      folders: values.length,
      failures,
      previousUnread: before ? before.unreadMessageCount : null,
      unread: after ? after.unreadMessageCount : null,
    };
  });

  tbxRegistry.define("folders.update", async (params) => {
    const folderId = tbxUtil.need(params, "folderId", "string");
    if (params.isFavorite === undefined || params.isFavorite === null) {
      throw tbxError.usage(
        "isFavorite is required — it is the only folder property this API can update"
      );
    }
    const before = await browser.folders.get(folderId);
    await browser.folders.update(folderId, { isFavorite: Boolean(params.isFavorite) });
    const after = await browser.folders.get(folderId);
    return { previous: summary(before, null), folder: summary(after, null) };
  });

  // ----------------------------------------------------------------- emptying out

  /* There is no folders.emptyTrash or folders.emptyJunk in the official API on 153,
   * so these do what the UI does: delete every message permanently, folder by folder,
   * and (for Trash) drop the emptied subfolders afterwards.
   *
   * Ids are collected before the first delete. Deleting invalidates the MessageList
   * we would otherwise still be paging through, and the symptom is a silent half-empty
   * folder rather than an error. */
  async function collectIds(folderId) {
    const ids = [];
    let current = await browser.messages.list(folderId);
    while (current) {
      for (const message of current.messages || []) {
        ids.push(message.id);
      }
      if (!current.id) {
        break;
      }
      current = await browser.messages.continueList(current.id);
      if (!current || !(current.messages || []).length) {
        break;
      }
    }
    return ids;
  }

  async function purge(folders, ctx, label) {
    const plan = [];
    for (const folder of folders) {
      for (const id of await collectIds(folder.id)) {
        plan.push(id);
      }
    }
    let deleted = 0;
    for (let offset = 0; offset < plan.length; offset += PURGE_CHUNK) {
      const chunk = plan.slice(offset, offset + PURGE_CHUNK);
      await browser.messages.delete(chunk, { deletePermanently: true });
      deleted += chunk.length;
      if (ctx && ctx.progress) {
        ctx.progress(deleted, plan.length, label);
      }
    }
    return deleted;
  }

  async function emptySpecial(params, ctx, use, removeSubFolders) {
    const folder = params.folderId
      ? await browser.folders.get(params.folderId)
      : await specialFolder(tbxUtil.need(params, "accountId", "string"), use, true);
    const children = (await browser.folders.getSubFolders(folder.id, true)) || [];
    const descendants = flatten(children, []);
    const deleted = await purge([folder, ...descendants], ctx, `emptying ${folder.name}`);

    let foldersRemoved = 0;
    if (removeSubFolders) {
      for (const child of children) {
        try {
          await browser.folders.delete(child.id);
          foldersRemoved += 1;
        } catch (ex) {
          // A folder the server refuses to drop is not worth failing the whole
          // operation over — its messages are already gone.
          tbxLog.warn(`could not remove ${child.path}: ${ex.message || ex}`);
        }
      }
    }
    return {
      folderId: folder.id,
      path: folder.path,
      accountId: folder.accountId,
      deleted,
      foldersRemoved,
      subFoldersEmptied: descendants.length,
    };
  }

  tbxRegistry.define("folders.emptyTrash", async (params, ctx) =>
    emptySpecial(params, ctx, "trash", params.removeSubFolders !== false)
  );

  // Junk keeps its structure: unlike Trash, its subfolders are usually filter
  // targets the user set up on purpose.
  tbxRegistry.define("folders.emptyJunk", async (params, ctx) =>
    emptySpecial(params, ctx, "junk", Boolean(params.removeSubFolders))
  );
}
